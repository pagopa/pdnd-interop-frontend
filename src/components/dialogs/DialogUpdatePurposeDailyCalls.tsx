import React from 'react'
import { Box, Button, Dialog, DialogActions, DialogContent, DialogTitle } from '@mui/material'
import { useTranslation } from 'react-i18next'
import { useDialog } from '@/stores'
import type { DialogUpdatePurposeDailyCallsProps } from '@/types/dialog.types'
import { TextField } from '../shared/ReactHookFormInputs'
import { PurposeMutations } from '@/api/purpose'
import { FormProvider, useForm } from 'react-hook-form'

type UpdateDailyCallsFormValues = { dailyCalls: number }

export const DialogUpdatePurposeDailyCalls: React.FC<DialogUpdatePurposeDailyCallsProps> = ({
  purposeId,
  dailyCalls = 1,
}) => {
  const ariaLabelId = React.useId()
  const { t } = useTranslation('shared-components', {
    keyPrefix: 'dialogUpdatePurposeDailyCalls',
  })

  const formMethods = useForm<UpdateDailyCallsFormValues>({
    defaultValues: {
      dailyCalls,
    },
  })

  const { closeDialog } = useDialog()
  const { mutate: updateDailyCalls } = PurposeMutations.useUpdateDailyCalls()

  const onSubmit = ({ dailyCalls }: UpdateDailyCallsFormValues) => {
    updateDailyCalls({ purposeId, dailyCalls }, { onSuccess: closeDialog })
  }

  return (
    <Dialog open onClose={closeDialog} aria-labelledby={ariaLabelId} fullWidth>
      <Box component="form" noValidate onSubmit={formMethods.handleSubmit(onSubmit)}>
        <DialogTitle id={ariaLabelId}>{t('title')}</DialogTitle>

        <DialogContent>
          <FormProvider {...formMethods}>
            <TextField
              type="number"
              name="dailyCalls"
              label={t('content.dailyCallsField.label')}
              infoLabel={t('content.dailyCallsField.infoLabel')}
              focusOnMount={true}
              inputProps={{ min: '1' }}
              rules={{ required: true, min: 1 }}
            />
          </FormProvider>
        </DialogContent>

        <DialogActions>
          <Button variant="outlined" onClick={closeDialog}>
            {t('actions.cancelLabel')}
          </Button>
          <Button variant="contained" type="submit">
            {t('actions.confirmLabel')}
          </Button>
        </DialogActions>
      </Box>
    </Dialog>
  )
}
