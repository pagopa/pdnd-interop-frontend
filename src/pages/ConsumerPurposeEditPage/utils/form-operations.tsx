import {
  CheckboxGroup,
  RadioGroup,
  Select,
  TextField,
} from '@/components/shared/ReactHookFormInputs'
import type { InputOption } from '@/types/common.types'
import React from 'react'
import identity from 'lodash/identity'
import type {
  Dependency,
  DynamicFormOperations,
  Questions,
  QuestionV1,
  QuestionV2,
} from '../types/risk-analysis.types'
import { RiskAnalysisSwitch } from '../components/PurposeEditStep2RiskAnalysis/RiskAnalysisSwitch'

export const dynamicFormOperationsVersions: DynamicFormOperations = {
  '1.0': {
    getUpdatedQuestions: (values, riskAnalysis) => {
      const updatedQuestions = riskAnalysis.questions.reduce((acc, next) => {
        // We are sure we can only queue the questions that come after the current one
        // because the dependencies are always on the questions that come before.
        // E.g. question 3 can only have question 1 and/or 2 as a dependency,
        // not question 4 and subsequent
        const queuedQuestions = Object.values(acc).map((d) => d.id)

        // Check the dependencies of this questions
        const satisfiesAllDependencies = next.dependencies.every((d) => {
          // If the question is a dependency
          const isInArray = queuedQuestions.includes(d.id)
          const id = d.id as string

          // and if the field currently has the same value as stated in the dependency
          const hasSameValue = Array.isArray(values[id])
            ? (values[id] as Array<unknown>).includes(d.value)
            : values[id] === d.value

          return isInArray && hasSameValue
        })

        // Then it needs to be added as a new question to ask
        if (satisfiesAllDependencies) {
          acc[next.id] = next
        }

        return acc
      }, {} as Questions)
      return updatedQuestions
    },

    buildForm: (questions, _, lang, t) => {
      const questionKeys = Object.keys(questions)

      const formComponents = questionKeys.map((id, i) => {
        const { type, label, options, infoLabel } = questions[id] as QuestionV1

        const sx = questionKeys.length - 1 === i ? { mb: 0 } : {}
        const inputOptions = options ? options.map((o) => ({ ...o, label: o.label[lang] })) : []
        const commonProps = {
          key: id,
          name: id,
          label: label[lang],
          infoLabel: infoLabel ? infoLabel[lang] : undefined,
          sx,
        }

        switch (type) {
          case 'text':
            return <TextField {...commonProps} rules={{ required: true }} />

          case 'select-one':
            return (
              <Select
                {...commonProps}
                options={inputOptions}
                emptyLabel={t('edit.step2.emptyLabel')}
                rules={{ required: true }}
              />
            )

          case 'checkbox':
            const errorMessage =
              options && options?.length > 1
                ? t('edit.step2.multiCheckboxField.validation.mixed.required')
                : t('edit.step2.singleCheckboxField.validation.mixed.required')
            return (
              <CheckboxGroup
                {...commonProps}
                options={inputOptions}
                rules={{
                  validate: (value) =>
                    (typeof value !== 'undefined' && value.length > 0) || errorMessage,
                }}
              />
            )

          case 'radio':
            return <RadioGroup {...commonProps} options={inputOptions} rules={{ required: true }} />
        }
      })

      return { formComponents: formComponents, isSubmitBtnDisabled: false }
    },
  },
  '2.0': {
    getUpdatedQuestions: (values, riskAnalysis) => {
      const questions = riskAnalysis.questions as Array<QuestionV2>

      // Filters all the questions that not satisfies the dependency inside the "dependencies" property
      const updatedQuestions = questions.filter(({ dependencies }) => {
        function satisfiesDependency(dep: QuestionV2['dependencies'][0]) {
          if (Array.isArray(values[dep.id])) {
            return (values[dep.id] as Array<unknown>).includes(dep.value)
          }
          if (Array.isArray(dep.value)) {
            return dep.value.includes(values[dep.id])
          }
          return values[dep.id] === dep.value
        }
        // If length of dependencies array is 0 the "every" method returns true
        return dependencies.every(satisfiesDependency)
      })

      // Array<Question> -> Record<"id", Question>
      return updatedQuestions.reduce((acc, next) => ({ ...acc, [next.id]: next }), {} as Questions)
    },

    buildForm: (questions, formMethods, lang, t) => {
      const values = formMethods.getValues()

      const formComponents: Array<React.ReactNode> = []

      const questionIds = Object.keys(questions)

      type QuestionOption = Exclude<QuestionV2['options'], undefined>[0]

      function checkOptionDependency(dependency: Dependency) {
        const currentDepValue = values[dependency.id]

        const hasDependency =
          currentDepValue === dependency.value ||
          (Array.isArray(currentDepValue) &&
            (currentDepValue as string[]).includes(dependency.value as string))

        return hasDependency
      }

      function parseOption(option: QuestionOption, { hideOption, id, defaultValue }: QuestionV2) {
        // if the key "hideOption" is present in the question object and the conditions are satisfied
        // the option will be not added to the array of options
        const shouldHideOption =
          hideOption &&
          hideOption[option.value] &&
          hideOption[option.value].some(checkOptionDependency)

        if (!shouldHideOption) {
          return { value: option.value, label: option.label[lang] }
        }

        if ((values[id] as string[])?.includes(option.value)) {
          formMethods.setValue(id, defaultValue)
        }
      }

      function buildFormQuestionComponents(
        { id, type, ...question }: QuestionV2,
        inputOptions: Array<InputOption>,
        isLast: boolean
      ) {
        const questionComponents: Array<React.ReactNode> = []

        const maxLength = question?.validation?.maxLength
        const isRequired = question.required
        const isMultipleChoice = question.dataType === 'multi'

        let label = question.label[lang]
        let infoLabel = question.infoLabel && question.infoLabel[lang]

        if (maxLength) {
          const maxLengthLabel = t('edit.step2.validation.maxLength', { num: maxLength })
          if (infoLabel) {
            infoLabel += `. ${maxLengthLabel}`
          } else {
            infoLabel = maxLengthLabel
          }
        }

        const labelValidation: Array<string> = []

        if (isRequired) {
          labelValidation.push(t('edit.step2.validation.required'))
        }

        if (isMultipleChoice) {
          labelValidation.push(t('edit.step2.validation.multipleChoice'))
        }

        if (labelValidation.length > 0) {
          label += ` (${labelValidation.join(', ')})`
        }

        const sx = isLast ? { mb: 0 } : {}
        const commonProps = { key: id, name: id, label, infoLabel, sx }

        switch (type) {
          case 'text':
            questionComponents.push(
              <TextField {...commonProps} inputProps={{ maxLength }} rules={{ required: true }} />
            )
            break
          case 'select-one':
            questionComponents.push(
              <Select
                {...commonProps}
                options={inputOptions}
                emptyLabel={t('edit.step2.emptyLabel')}
                rules={{ required: true }}
              />
            )
            break
          case 'checkbox':
            questionComponents.push(
              <CheckboxGroup
                {...commonProps}
                options={inputOptions}
                rules={{
                  validate: (value) =>
                    (typeof value !== 'undefined' && value.length > 0) ||
                    t('edit.step2.multiCheckboxField.validation.mixed.required'),
                }}
              />
            )
            break
          case 'radio':
            questionComponents.push(
              <RadioGroup {...commonProps} options={inputOptions} rules={{ required: true }} />
            )
            break
          case 'switch':
            questionComponents.push(
              <RiskAnalysisSwitch
                {...commonProps}
                options={inputOptions}
                rules={{
                  validate: (value) =>
                    (typeof value === 'boolean' && value === true) ||
                    t('edit.step2.riskAnalysisSwitch.validation.boolean.isValue'),
                }}
              />
            )
            break
        }

        return questionComponents
      }

      for (let i = 0; i < questionIds.length; i++) {
        const questionId = questionIds[i]

        const question = questions[questionId] as QuestionV2

        const inputOptions = (question.options
          ?.map((option) => parseOption(option, question))
          .filter(identity) || []) as Array<InputOption>

        formComponents.push(
          ...buildFormQuestionComponents(question, inputOptions, questionIds.length - 1 === i)
        )
      }

      return { formComponents, isSubmitBtnDisabled: false }
    },
  },
}

export function getFormOperations(version: string) {
  const dynamicFormOperations = dynamicFormOperationsVersions[version]

  if (!dynamicFormOperations) {
    throw new Error('There is no business logic for this specific risk analysis data form')
  }

  return dynamicFormOperations
}
