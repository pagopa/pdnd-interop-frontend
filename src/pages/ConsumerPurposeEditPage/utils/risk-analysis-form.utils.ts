import type { Dependency, FormConfigQuestion, HideOption } from '@/api/api.generatedTypes'
import type { AnswerValue, Answers, Questions } from '../types/risk-analysis-form.types'

/**
 * Transform the answers from the backend to the frontend format
 * The answer from the backend arrives as an array of strings, but the frontend
 * needs a specific format depending on the `visualType` value.
 *
 * - `switch` needs a boolean
 * - `checkbox` needs an array of strings
 * - the rest needs a string
 *
 * @param backendAnswer The answer from the backend
 * @param question The question that the answer belongs to
 * @returns The answer in the type that the frontend needs
 */
export function getFrontendAnswerValue(
  backendAnswer: Array<string>,
  visualType?: FormConfigQuestion['visualType']
): AnswerValue {
  if (visualType === 'switch') {
    return backendAnswer[0] === 'true'
  }

  if (visualType === 'checkbox') {
    return backendAnswer
  }

  return backendAnswer[0]
}

/**
 * Transform the answers from the frontend to the backend format
 * The backend always needs an array of strings.
 *
 * @param answer The answer from the frontend
 * @returns The answer in the type that the backend needs
 * */
export function getBackendAnswerValue(answer: AnswerValue): Array<string> {
  if (typeof answer === 'boolean') {
    return [answer.toString()]
  }

  if (Array.isArray(answer)) {
    return answer
  }

  if (typeof answer === 'string') {
    return [answer]
  }

  throw new Error(`Invalid answer type: ${typeof answer} \n\n ${JSON.stringify(answer, null, 2)}`)
}

/**
 * Returns the valid answers that should be sent to the backend.
 * The answers that should be sent to the backend are only the answers
 * to the questions that are currently visible to the user.
 * The answers to the questions that are not visible to the user are not sent.
 *
 * @param currentQuestionsIds - the ids of the questions that are currently visible to the user
 * @param answers - the answers to all the questions
 * @returns the valid answers that should be sent to the backend
 * */
export function getValidAnswers(currentQuestionsIds: Array<string>, answers: Answers) {
  return Object.keys(answers).reduce<Record<string, Array<string>>>((acc, key) => {
    // If the question is actually visible, add it to the valid answers
    if (currentQuestionsIds.includes(key)) {
      return { ...acc, [key]: getBackendAnswerValue(answers[key]) }
    }
    return acc
  }, {})
}

/**
 * Returns true if the dependency is satisfied, false otherwise.
 *
 * @param dependency - the dependency
 * @param answers - the actual form values
 * @returns `true` if the dependency is satisfied, `false` otherwise
 * */
export function isDependencySatisfied(dependency: Dependency | HideOption, answers: Answers) {
  const answer = answers[dependency.id]
  if (Array.isArray(answer)) {
    return answer.includes(dependency.value)
  }

  return answer === dependency.value
}

/**
 * Returns the updated question data.
 *
 * @param answers - the actual form values
 * @param riskAnalysis - the risk analysis document
 *
 * @returns the updated questions data
 * */
export function getUpdatedQuestions(
  answers: Answers,
  riskAnalysisQuestions: FormConfigQuestion[]
): Questions {
  // Filters all the questions that not satisfies the dependency inside the "dependencies" property
  return riskAnalysisQuestions.reduce<Questions>((acc, question) => {
    const doesSatisfyDeps = question.dependencies.every((dependency) =>
      isDependencySatisfied(dependency, answers)
    )
    if (doesSatisfyDeps) {
      acc[question.id] = question
    }
    return acc
  }, {})
}

/**
 * Returns the default values for the risk analysis form.
 * The default values are the values that the form should have when it is first loaded.
 * The default values are the values that the user has already answered, if any.
 * If the user has not answered any question, the values are the default values
 * defined in the risk analysis document.
 *
 * @param riskAnalysisConfig the risk analysis document
 * @param backendAnswers the answers data, taken from the Purpose
 * @returns the default values for the risk analysis form
 */
export function getRiskAnalysisDefaultValues(
  riskAnalysisConfigQuestions: FormConfigQuestion[],
  backendAnswers?: Record<string, Array<string>>
): Answers {
  return riskAnalysisConfigQuestions.reduce<Answers>((acc, question) => {
    const answer = backendAnswers?.[question.id] ?? question.defaultValue
    acc[question.id] = getFrontendAnswerValue(answer, question.visualType)

    if (!acc[question.id] && question.dataType === 'FREETEXT') {
      acc[question.id] = ''
    }

    return acc
  }, {})
}
