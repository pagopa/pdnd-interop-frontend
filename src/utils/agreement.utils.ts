import type { AgreementSummary } from '@/types/agreement.types'
import type { ProviderOrConsumer } from '@/types/common.types'

export const canAgreementBeUpgraded = (agreement: AgreementSummary, mode: ProviderOrConsumer) => {
  if (!agreement || mode !== 'consumer') return false

  return !!(
    agreement.eservice.activeDescriptor &&
    agreement.eservice.activeDescriptor.state === 'PUBLISHED' &&
    agreement.eservice.activeDescriptor.version > agreement.eservice.version &&
    agreement.state !== 'ARCHIVED'
  )
}
