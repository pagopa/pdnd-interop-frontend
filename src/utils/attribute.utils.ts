import type { FrontendAttributes, PartyAttribute } from './../types/attribute.types'
import has from 'lodash/has'
import type { EServiceAttribute, EServiceAttributes } from '@/api/api.generatedTypes'
import identity from 'lodash/identity'

export function checkEServiceAttributesOwnership(
  partyAttributes: Array<PartyAttribute>,
  eserviceAttributes: Array<EServiceAttribute>
) {
  const partyAttributesIds = partyAttributes.filter((p) => p.state === 'ACTIVE').map((p) => p.id)

  const hasAllAttributes = eserviceAttributes.every((eserviceAttribute) => {
    if (has(eserviceAttribute, 'single')) {
      const match = partyAttributesIds.find((id) => id === eserviceAttribute.single?.id)
      return Boolean(match)
    }

    const match = partyAttributesIds.find((id) => {
      const groupIds = eserviceAttribute.group?.map((a) => a.id)
      return groupIds?.includes(id)
    })
    return Boolean(match)
  })

  return hasAllAttributes
}

export function remapEServiceAttributes(backendAttributes: EServiceAttributes): FrontendAttributes {
  const mappedAttributes: FrontendAttributes = Object.keys(backendAttributes).reduce(
    (acc, attributeType) => {
      const mapped = backendAttributes[attributeType as keyof EServiceAttributes]
        .map((attribute: EServiceAttribute) => {
          const isSingle = has(attribute, 'single')

          if (isSingle) {
            const single = attribute.single!
            const { id, explicitAttributeVerification, description, name } = single
            return {
              attributes: [{ id, description, name, kind: attributeType.toUpperCase() }],
              explicitAttributeVerification,
            }
          }

          const isGroup = has(attribute, 'group')

          if (isGroup) {
            const group = attribute.group
            return {
              attributes: [
                ...group!.map(({ id, description, name }) => ({
                  id,
                  description,
                  name,
                  kind: attributeType.toUpperCase(),
                })),
              ],
              explicitAttributeVerification: group![0].explicitAttributeVerification,
            }
          }
        })
        .filter(identity)
      return { ...acc, [attributeType]: mapped }
    },
    { certified: [], verified: [], declared: [] }
  )

  return mappedAttributes
}
