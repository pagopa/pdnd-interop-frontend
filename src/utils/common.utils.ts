import type { TopSideActions } from '@/components/layout/containers/PageContainer'
import type { ActionItem } from '@/types/common.types'
import type { ButtonProps } from '@mui/material'
import React from 'react'

/**
 * Top side actions are formatted with the first action as a button, and the
 * rest inside of an actionMenu.
 */
export function formatTopSideActions(
  actions: Array<ActionItem>,
  buttonProps?: Omit<ButtonProps, keyof ActionItem | 'onClick'>
): TopSideActions | undefined {
  return actions.length > 0
    ? {
        buttons: [{ ...actions[0], ...buttonProps }],
        actionMenu: actions.slice(1).length > 0 ? actions.slice(1) : undefined,
      }
    : undefined
}

export async function waitFor(ms: number) {
  return new Promise((resolve) => setTimeout(resolve, ms))
}

export function createContext<ContextValue>(name: string, defaultValue: ContextValue) {
  const context = React.createContext<ContextValue>(defaultValue)
  context.displayName = name
  const useContext = () => React.useContext<ContextValue>(context)
  return {
    useContext,
    Provider: context.Provider,
  } as const
}
