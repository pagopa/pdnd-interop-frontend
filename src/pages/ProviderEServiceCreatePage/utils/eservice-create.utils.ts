import type {
  AgreementApprovalPolicy,
  DescriptorAttributes,
  ProducerEServiceDescriptor,
} from '@/api/api.generatedTypes'
import type { RemappedEServiceAttributes } from '@/types/attribute.types'
import { getKeys } from '@/utils/array.utils'
import { remapEServiceAttributes } from '@/utils/attribute.utils'
import isEqual from 'lodash/isEqual'

type DescriptorDataStep2ToCompare = {
  voucherLifespan: number
  audience: string[]
  agreementApprovalPolicy: AgreementApprovalPolicy
  version: string
  description: string
  dailyCallsPerConsumer: number
  dailyCallsTotal: number
}

export function compareDescriptorsStep2DataIfEquals(
  newDescriptorData: DescriptorDataStep2ToCompare,
  descriptorToCompare: ProducerEServiceDescriptor
) {
  const descriptorDataToCompare: DescriptorDataStep2ToCompare = {
    voucherLifespan: descriptorToCompare.voucherLifespan,
    audience: descriptorToCompare.audience,
    agreementApprovalPolicy: descriptorToCompare.agreementApprovalPolicy,
    version: descriptorToCompare.version,
    description: descriptorToCompare.description ?? '',
    dailyCallsPerConsumer: descriptorToCompare.dailyCallsPerConsumer,
    dailyCallsTotal: descriptorToCompare.dailyCallsTotal,
  }

  return getKeys(newDescriptorData).every((key) =>
    isEqual(newDescriptorData[key], descriptorDataToCompare[key])
  )
}

export function compareAttributesStep3IfEquals(
  newAttributes: RemappedEServiceAttributes,
  attributesToCompare: DescriptorAttributes
) {
  const remappedAttributesToCompare = remapEServiceAttributes(attributesToCompare)

  const newAttributesToCompare = {
    certified: newAttributes.certified.filter((group) => group.attributes.length > 0),
    verified: newAttributes.verified.filter((group) => group.attributes.length > 0),
    declared: newAttributes.declared.filter((group) => group.attributes.length > 0),
  }

  return getKeys(newAttributes).every((key) =>
    isEqual(newAttributesToCompare[key], remappedAttributesToCompare[key])
  )
}
