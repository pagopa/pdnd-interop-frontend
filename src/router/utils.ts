import { routes } from '@/router/routes'
import type { LangCode, ProviderOrConsumer } from '@/types/common.types'
import type { RouteConfig, RouteKey } from '@/router/types'
import { generatePath, matchPath } from 'react-router-dom'
import { getKeys } from '@/utils/array.utils'
import memoize from 'lodash/memoize'
import identity from 'lodash/identity'
import sortBy from 'lodash/sortBy'
import { PUBLIC_URL } from '@/config/env'
import qs from 'qs'
import QueryString from 'qs'

/** Returns the localized path of the given routeKey and language  */
export function getLocalizedRoutePathname(routeKey: RouteKey, lang: LangCode) {
  return `${PUBLIC_URL}/${lang}/${routes[routeKey].PATH[lang]}`
}

/** Checks if the routeKey represent the given pathname */
function matchRouteKeyPath(pathname: string, lang: LangCode, routeKey: RouteKey) {
  return matchPath(getLocalizedRoutePathname(routeKey as RouteKey, lang), pathname)
}

/** Returns the routeKey of the given pathname */
export const getRouteKeyFromPathname = memoize(function _getRouteKeyFromPathname(
  pathname: string,
  lang: LangCode
) {
  let pathnameWithBaseUrl = pathname
  if (!pathnameWithBaseUrl.startsWith(PUBLIC_URL)) {
    pathnameWithBaseUrl = PUBLIC_URL + pathnameWithBaseUrl
  }
  const isTheRouteKeyPath = matchRouteKeyPath.bind(null, pathnameWithBaseUrl, lang)
  const routeKey = getKeys(routes).find(isTheRouteKeyPath)

  if (!routeKey) {
    throw new Error(`Pathname ${pathnameWithBaseUrl} has no associated routeKey.`)
  }

  return routeKey
})

export function switchPathLang(fromLang: LangCode, toLang: LangCode) {
  for (const routeKey of Object.keys(routes)) {
    const match = matchRouteKeyPath(window.location.pathname, fromLang, routeKey as RouteKey)

    if (match) {
      let newPath = generatePath(
        getLocalizedRoutePathname(routeKey as RouteKey, toLang),
        match.params as Record<string, string>
      )

      if (window.location.search) {
        newPath += window.location.search
      }

      if (window.location.hash) {
        newPath += window.location.hash
      }

      window.location.replace(newPath)
      return
    }
  }
}

export const URL_FRAGMENTS: Record<string, Record<LangCode, string>> = {
  FIRST_DRAFT: { it: 'prima-bozza', en: 'first-draft' },
  EDIT: { it: 'modifica', en: 'bozza' },
}

export function getSplittedPath(route: RouteConfig, currentLang: LangCode) {
  return route.PATH[currentLang].split('/').filter(identity)
}

export const getParentRoutes = memoize((routeKey: RouteKey): Array<RouteKey> => {
  function isParentRoute(
    possibleParentRouteSubpaths: Array<string>,
    currentSubpaths: Array<string>
  ) {
    if (possibleParentRouteSubpaths.length >= currentSubpaths.length) {
      return false
    }

    const allSameFragments = possibleParentRouteSubpaths.every(
      (pathFragment, i) => pathFragment === currentSubpaths[i]
    )

    // URL_FRAGMENTS.EDIT is appended at the end of a read path of the same type.
    // E.g. /eservice/myid becomes /eservice/myid/edit. So it's always same length + 1
    const isFalseParent = currentSubpaths.length === possibleParentRouteSubpaths.length + 1

    if (allSameFragments && isEditPath(routeKey) && isFalseParent) {
      return false
    }

    return allSameFragments
  }

  const route = routes[routeKey]
  const currentSubpaths = getSplittedPath(route, 'it')

  const parents = Object.entries(routes).filter(([_, possibleParentRoute]) =>
    isParentRoute(getSplittedPath(possibleParentRoute, 'it'), currentSubpaths)
  )

  const sortedParents = sortBy(parents, ([_, parent]) => getSplittedPath(parent, 'it').length)

  return sortedParents.map(([routeKey]) => routeKey) as Array<RouteKey>
})

export const isProviderOrConsumerRoute = memoize(
  (routeKey: RouteKey): ProviderOrConsumer | null => {
    const excludeList = ['ui', 'it']
    const subroutes = getSplittedPath(routes[routeKey], 'it').filter(
      (b) => !excludeList.includes(b)
    )
    const [mode] = subroutes

    if (mode === 'erogazione') {
      return 'provider'
    }

    if (mode === 'fruizione') {
      return 'consumer'
    }

    return null
  }
)

export const isEditPath = memoize((routeKey: RouteKey): boolean => {
  const subroutes = getSplittedPath(routes[routeKey], 'it')

  const lastBit = subroutes[subroutes.length - 1]
  return Object.values(URL_FRAGMENTS.EDIT).some((f) => lastBit.endsWith(f))
})

export function parseSearch(search: string) {
  return qs.parse(search, { ignoreQueryPrefix: true })
}

export function stringifySearch(searchObj: Record<string, string> | QueryString.ParsedQs) {
  return qs.stringify(searchObj)
}

/**
 * Returns an array with all the dynamic path names for a given RouteKey
 * @example
 * "/:foo/test/:bar" => ["foo", "bar"]
 */
export const getDynamicPathsName = memoize((routeKey: RouteKey) => {
  return routes[routeKey].PATH.it
    .split('/')
    .filter(identity)
    .filter((subpath) => subpath.startsWith(':'))
    .map((param) => param.replace(':', ''))
})
