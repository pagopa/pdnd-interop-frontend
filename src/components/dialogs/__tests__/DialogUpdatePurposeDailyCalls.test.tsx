import React from 'react'
import { DialogUpdatePurposeDailyCalls } from '../DialogUpdatePurposeDailyCalls'
import { renderWithApplicationContext } from '@/utils/testing.utils'

describe('DialogUpdatePurposeDailyCalls testing', () => {
  it('should match the snapshot', () => {
    const screen = renderWithApplicationContext(
      <DialogUpdatePurposeDailyCalls type="updatePurposeDailyCalls" purposeId="purposeId" />,
      {
        withReactQueryContext: true,
      }
    )
    expect(screen.baseElement).toMatchSnapshot()
  })
})
