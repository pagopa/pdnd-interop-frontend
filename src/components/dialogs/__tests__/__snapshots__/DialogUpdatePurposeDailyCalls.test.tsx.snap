// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DialogUpdatePurposeDailyCalls testing > should match the snapshot 1`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="MuiDialog-root MuiModal-root css-zw3mfo-MuiModal-root-MuiDialog-root"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root MuiModal-backdrop css-yiavyu-MuiBackdrop-root-MuiDialog-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-testid="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiDialog-container MuiDialog-scrollPaper css-hz1bth-MuiDialog-container"
      role="presentation"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      tabindex="-1"
    >
      <div
        aria-labelledby=":r0:"
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm MuiDialog-paperFullWidth css-tlc64q-MuiPaper-root-MuiDialog-paper"
        role="dialog"
      >
        <form
          class="MuiBox-root css-0"
          novalidate=""
        >
          <h2
            class="MuiTypography-root MuiTypography-h6 MuiDialogTitle-root css-bdhsul-MuiTypography-root-MuiDialogTitle-root"
            id=":r0:"
          >
            title
          </h2>
          <div
            class="MuiDialogContent-root css-ypiqx9-MuiDialogContent-root"
          >
            <div
              class="MuiBox-root css-zitybv"
            >
              <div
                class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled Mui-focused MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="true"
                  for=":r3:"
                  id=":r3:-label"
                >
                  content.dailyCallsField.label
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-focused MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                    id=":r3:"
                    min="1"
                    name="dailyCalls"
                    type="number"
                    value="1"
                  />
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-14lo706"
                    >
                      <span>
                        content.dailyCallsField.label
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
              <p
                class="MuiTypography-root MuiTypography-caption css-6n3pnq-MuiTypography-root"
              >
                content.dailyCallsField.infoLabel
              </p>
            </div>
          </div>
          <div
            class="MuiDialogActions-root MuiDialogActions-spacing css-hlj6pa-MuiDialogActions-root"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              actions.cancelLabel
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="submit"
            >
              actions.confirmLabel
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </form>
      </div>
    </div>
    <div
      data-testid="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
